File: smarthomec++/EnergyReport.cpp
#include "energyReport.hpp"
#include "device.hpp"
#include <iostream>
#include <algorithm>
#include <vector>
#include <iomanip>
#include <sstream>

EnergyReport::EnergyReport(const std::string& id, std::time_t start, std::time_t end)
    : reportId(id), periodStart(start), periodEnd(end),
    totalConsumption(0.0), peakLoad(0.0) {
}

EnergyReport::~EnergyReport() {
    deviceConsumptions.clear();
}

void EnergyReport::generateReport() {
    totalConsumption = 0.0;
    peakLoad = 0.0;

    for (const auto& pair : deviceConsumptions) {
        totalConsumption += pair.second;
        if (pair.second > peakLoad) {
            peakLoad = pair.second;
        }
    }
}

std::vector<std::shared_ptr<Device>> EnergyReport::getTopConsumingDevices(int count) const {
    std::vector<std::pair<std::shared_ptr<Device>, double>> devices;
    for (const auto& pair : deviceConsumptions) {
        devices.push_back(pair);
    }

    std::sort(devices.begin(), devices.end(),
        [](const auto& a, const auto& b) { return a.second > b.second; });

    std::vector<std::shared_ptr<Device>> topDevices;
    for (int i = 0; i < std::min(count, static_cast<int>(devices.size())); ++i) {
        topDevices.push_back(devices[i].first);
    }

    return topDevices;
}

void EnergyReport::addDeviceConsumption(std::shared_ptr<Device> device, double consumption) {
    deviceConsumptions[device] = consumption;
}

void EnergyReport::displayReport() const {
    std::cout << "\n=== ќ“„≈“ ѕќ ЁЌ≈–√ќѕќ“–≈ЅЋ≈Ќ»ё " << reportId << " ===" << std::endl;

    char startBuffer[80], endBuffer[80];

    // Ѕезопасна€ верси€ дл€ Windows
#ifdef _WIN32
    struct tm startInfo, endInfo;
    localtime_s(&startInfo, &periodStart);
    localtime_s(&endInfo, &periodEnd);
    std::strftime(startBuffer, sizeof(startBuffer), "%Y-%m-%d %H:%M:%S", &startInfo);
    std::strftime(endBuffer, sizeof(endBuffer), "%Y-%m-%d %H:%M:%S", &endInfo);
#else
    struct tm* startInfo = std::localtime(&periodStart);
    struct tm* endInfo = std::localtime(&periodEnd);
    std::strftime(startBuffer, sizeof(startBuffer), "%Y-%m-%d %H:%M:%S", startInfo);
    std::strftime(endBuffer, sizeof(endBuffer), "%Y-%m-%d %H:%M:%S", endInfo);
#endif

    std::cout << "ѕериод: с " << startBuffer << " по " << endBuffer << std::endl;
    std::cout << "ќбщее потребление: " << std::fixed << std::setprecision(2)
        << totalConsumption << " к¬тЈч" << std::endl;
    std::cout << "ѕикова€ нагрузка: " << peakLoad << " к¬т" << std::endl;
    std::cout << "¬сего устройств в отчете: " << deviceConsumptions.size() << std::endl;

    auto topDevices = getTopConsumingDevices(3);
    if (!topDevices.empty()) {
        std::cout << "\n“оп-3 потребителей энергии:" << std::endl;
        for (size_t i = 0; i < topDevices.size(); i++) {
            const auto& device = topDevices[i];
            double consumption = deviceConsumptions.at(device);
            std::cout << "  " << i + 1 << ". " << device->getName()
                << " (" << device->getDeviceTypeString() << "): "
                << consumption << " к¬тЈч" << std::endl;
        }
    }

    std::cout << "\nƒетализаци€ по устройствам:" << std::endl;
    int counter = 1;
    for (const auto& pair : deviceConsumptions) {
        std::cout << "  " << counter++ << ". " << pair.first->getName()
            << ": " << pair.second << " к¬тЈч" << std::endl;
    }

    double averageLoad = deviceConsumptions.empty() ? 0.0 : totalConsumption / deviceConsumptions.size();
    std::cout << "\n—редн€€ нагрузка на устройство: " << averageLoad << " к¬тЈч" << std::endl;

    std::cout << "======================================" << std::endl;
}

std::string EnergyReport::getReportId() const {
    return reportId;
}

double EnergyReport::getTotalConsumption() const {
    return totalConsumption;
}

double EnergyReport::getPeakLoad() const {
    return peakLoad;
}

std::string EnergyReport::serialize() const {
    std::stringstream ss;
    ss << reportId << "|"
        << periodStart << "|"
        << periodEnd << "|"
        << totalConsumption << "|"
        << peakLoad << "|"
        << deviceConsumptions.size();

    for (const auto& pair : deviceConsumptions) {
        ss << "|" << pair.first->getId() << ":" << pair.second;
    }

    return ss.str();
}

std::unique_ptr<EnergyReport> EnergyReport::deserialize(const std::string& data) {
    try {
        std::stringstream ss(data);
        std::string id, startStr, endStr, totalStr, peakStr, countStr;

        std::getline(ss, id, '|');
        std::getline(ss, startStr, '|');
        std::getline(ss, endStr, '|');
        std::getline(ss, totalStr, '|');
        std::getline(ss, peakStr, '|');
        std::getline(ss, countStr, '|');

        auto report = std::make_unique<EnergyReport>(
            id,
            std::stol(startStr),
            std::stol(endStr)
        );

        report->totalConsumption = std::stod(totalStr);
        report->peakLoad = std::stod(peakStr);

        int deviceCount = std::stoi(countStr);

        // ѕримечание: устройства должны быть добавлены позже
        // после загрузки всех данных

        return report;
    }
    catch (const std::exception& e) {
        std::cerr << "ќшибка десериализации энергоотчета: " << e.what() << std::endl;
        return nullptr;
    }
}

File: smarthomec#/EnergyReport.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SmartHomeSystem
{
    public class EnergyReport
    {
        public string ReportId { get; private set; }
        public DateTime PeriodStart { get; set; }
        public DateTime PeriodEnd { get; set; }
        public double TotalConsumption { get; private set; }
        public double PeakLoad { get; private set; }
        public Dictionary<Device, double> DeviceConsumptions { get; private set; }

        public EnergyReport(string reportId, DateTime periodStart, DateTime periodEnd)
        {
            ReportId = reportId;
            PeriodStart = periodStart;
            PeriodEnd = periodEnd;
            DeviceConsumptions = new Dictionary<Device, double>();
        }

        public void AddDeviceConsumption(Device device, double consumption)
        {
            DeviceConsumptions[device] = consumption;
        }

        public void GenerateReport()
        {
            TotalConsumption = DeviceConsumptions.Values.Sum();
            PeakLoad = DeviceConsumptions.Values.DefaultIfEmpty(0).Max();
        }

        public List<Device> GetTopConsumingDevices(int topN)
        {
            return DeviceConsumptions
                .OrderByDescending(kv => kv.Value)
                .Take(topN)
                .Select(kv => kv.Key)
                .ToList();
        }

        public void DisplayReport()
        {
            Console.WriteLine("\n══════════════════════════════════════════════");
            Console.WriteLine("               ОТЧЕТ ПО ЭНЕРГОПОТРЕБЛЕНИЮ");
            Console.WriteLine($"               {ReportId}");
            Console.WriteLine("══════════════════════════════════════════════");
            Console.WriteLine($"Период: {PeriodStart:dd.MM.yyyy} - {PeriodEnd:dd.MM.yyyy}");
            Console.WriteLine($"Общее потребление: {TotalConsumption:F2} кВт·ч");
            Console.WriteLine($"Пиковая нагрузка: {PeakLoad:F2} кВт");
            Console.WriteLine("\nПотребление по устройствам:");
            Console.WriteLine("┌────┬────────────────────┬─────────────────────┐");
            Console.WriteLine("│ №  │ Устройство         │ Потребление (кВт·ч)  │");
            Console.WriteLine("├────┼────────────────────┼─────────────────────┤");

            int counter = 1;
            foreach (var kvp in DeviceConsumptions.OrderByDescending(kv => kv.Value))
            {
                Console.WriteLine($"│ {counter,2} │ {kvp.Key.Name,-20} │ {kvp.Value,20:F2} │");
                counter++;
            }

            Console.WriteLine("└────┴────────────────────┴─────────────────────┘");
            Console.WriteLine("══════════════════════════════════════════════\n");
        }

        public string Serialize()
        {
            return $"{ReportId}|{PeriodStart}|{PeriodEnd}|{TotalConsumption}|{PeakLoad}";
        }

        public static EnergyReport Deserialize(string data)
        {
            var parts = data.Split('|');
            if (parts.Length != 5) return null;

            return new EnergyReport(
                parts[0],
                DateTime.Parse(parts[1]),
                DateTime.Parse(parts[2])
            )
            {
                TotalConsumption = double.Parse(parts[3]),
                PeakLoad = double.Parse(parts[4])
            };
        }

        public override string ToString()
        {
            return $"{ReportId} - {TotalConsumption:F2} кВт·ч";
        }
    }
}